src/data_c.c

```diff
@@ -324,1073 +324,201 @@
-
-    shmem_internal_get(target, source, sizeof(long double) * len, pe);
-    shmem_internal_get_wait();
-}
-
-
-
-void
-shmem_char_get(char *target, const char *source, size_t len, int pe)
-{
-    SHMEM_ERR_CHECK_INITIALIZED();
-
-    shmem_internal_get(target, source, sizeof(char) * len, pe);
-    shmem_internal_get_wait();
-}
-
-
-
-void
-shmem_short_get(short *target, const short *source, size_t len, int pe)
-{
-    SHMEM_ERR_CHECK_INITIALIZED();
-
-    shmem_internal_get(target, source, sizeof(short) * len, pe);
-    shmem_internal_get_wait();
-}
-
-
-void
-shmem_int_get(int *target, const int *source, size_t len, int pe)
-{
-    SHMEM_ERR_CHECK_INITIALIZED();
-
-    shmem_internal_get(target, source, sizeof(int) * len, pe);
-    shmem_internal_get_wait();
-}
-
-
-void
-shmem_long_get(long *target, const long *source, size_t len, int pe)
-{
-    SHMEM_ERR_CHECK_INITIALIZED();
-
-    shmem_internal_get(target, source, sizeof(long) * len, pe);
-    shmem_internal_get_wait();
-}
-
-
-void
-shmem_longlong_get(long long *target, const long long *source, size_t len, int pe)
-{
-    SHMEM_ERR_CHECK_INITIALIZED();
-
-    shmem_internal_get(target, source, sizeof(long long) * len, pe);
-    shmem_internal_get_wait();
-}
-
-
-void
-shmem_get8(void *dest, const void *source, size_t nelems, int pe)
-{
-  SHMEM_ERR_CHECK_INITIALIZED();
-
-  shmem_internal_get(dest, source, nelems, pe);
-  shmem_internal_get_wait();
-}
-
-
-void
-shmem_get16(void *dest, const void *source, size_t nelems, int pe)
-{
-  SHMEM_ERR_CHECK_INITIALIZED();
-
-  shmem_internal_get(dest, source, 2 * nelems, pe);
-  shmem_internal_get_wait();
-}
-
-
-void
-shmem_get32(void *target, const void *source, size_t len, int pe)
-{
-    SHMEM_ERR_CHECK_INITIALIZED();
-
-    shmem_internal_get(target, source, 4 * len, pe);
-    shmem_internal_get_wait();
-}
-
-
-void
-shmem_get64(void *target, const void *source, size_t len, int pe)
-{
-    SHMEM_ERR_CHECK_INITIALIZED();
-
-    shmem_internal_get(target, source, 8 * len, pe);
-    shmem_internal_get_wait();
-}
-
-
-void
-shmem_get128(void *target, const void *source, size_t len, int pe)
-{
-    SHMEM_ERR_CHECK_INITIALIZED();
-
-    shmem_internal_get(target, source, 16 * len, pe);
-    shmem_internal_get_wait();
-}
-
-
-void
-shmem_getmem(void *target, const void *source, size_t len, int pe)
-{
-    SHMEM_ERR_CHECK_INITIALIZED();
-
-    shmem_internal_get(target, source, len, pe);
+#define SHMEM_DEF_P(STYPE,TYPE) \
+  void shmem_##STYPE##_p(TYPE *addr, TYPE value, int pe)       \
+  {                                                            \
+    SHMEM_ERR_CHECK_INITIALIZED();                             \
+    shmem_internal_put_small(addr, &value, sizeof(value), pe); \
+  }
+
+SHMEM_DEFINE_FOR_RMA(SHMEM_DEF_P)
+
+#define SHMEM_DEF_G(STYPE,TYPE) \
+  TYPE shmem_##STYPE##_g(const TYPE *addr, int pe)   \
+  {                                                  \
+    TYPE tmp;                                        \
```

Comment:
I think uninitialized might be slightly better here, since there is a remote chance the compiler will report usage of an uninitialized variable if we screw up somewhere.